docs/proposals/volume-provisioning.md

```diff
@@ -126,30 +129,247 @@ We propose that:
 
-Existing behavior is un-changed for claims that do not specify `claim.Spec.Class`.
+  3. It tries to find a StorageClass instance referenced by annotation
+     `claim.Annotations["volume.beta.kubernetes.io/storage-class"]`. If not
+     found, it SHOULD report an error (by sending an event to the claim) and it
+     SHOULD retry periodically with step i.
+
+  4. The provisioner MUST parse arguments in the `StorageClass` and
+     `claim.Spec.Selector` and provisions appropriate storage asset that matches
+     both the parameters and the selector.
+     When it encounters unknown parameters in `storageClass.Parameters` or
+     `claim.Spec.Selector` or the combination of these parameters is impossible
+     to achieve, it SHOULD report an error and it MUST NOT provision a volume.
+     All errors found during parsing or provisioning SHOULD be send as events
+     on the claim and the provisioner SHOULD retry periodically with step i.
+
+  5. When the volume is provisioned, the provisioner MUST create a new PV
+     representing  the storage asset and save it in Kubernetes. When this fails,
+     it SHOULD retry creating the PV again few times. If all attempts fail, it
+     MUST delete the storage asset. All errors SHOULD be sent as events to the
+     claim.
+
+     The created PV MUST have these properties:
+
+     * `pv.Spec.ClaimRef` MUST point to the claim that led to its creation
+       (including the claim UID).
+
+       *This way, the PV will be bound to the claim.*
+
+     * `pv.Annotations["pv.kubernetes.io/provisioned-by"]` MUST be set to name
+       of the external provisioner. This provisioner will be used to delete the
+       volume.
+
+       *The provisioner/delete should not assume there is any other
+       provisioner/deleter available that would delete the volume.*
+
+     * `pv.Annotations["volume.beta.kubernetes.io/storage-class"]` MUST be set
+       to name of the storage class requested by the claim.
+
+       *So the created PV matches the claim.*
+
+     * The provisioner MAY store any other information to the created PV as
+       annotations. It SHOULD save any information that is needed to delete the
+       storage asset there, as appropriate StorageClass instance may not exist
+       when the volume will be deleted.
+
+     * `pv.Labels` MUST be set to match `claim.spec.selector`. The provisioner
+       MAY add additional labels.
+
+       *So the created PV matches the claim.*
+
+     * `pv.Spec` MUST be set to match requirements in `claim.Spec`, especially
+       access mode and PV size. The provisioned volume size MUST NOT be smaller
+       than size requested in the claim, however it MAY be larger.
+
+       *So the created PV matches the claim.*
+
+     * `pv.Spec.PersistentVolumeSource` MUST be set to point to the created
+       storage asset.
+
+     * `pv.Spec.PersistentVolumeReclaimPolicy` SHOULD be set to `Delete` unless
+       user manually configures other reclaim policy.
+
```

Comment:
Should there be an explicit mechanism for doing this that all external provisioners should support, like an annotation in the PVC?  Should this also apply to internal provisioners?